### DISCUSSION TITLE
关于任务流程数据组织形式的想法

### DISCUSSION THREAD

**Discussion by AsterNighT** [MaaAssistantArknights/MaaAssistantArknights#548] [Ideas]
目前 `resource/tasks.json` 这个文件有接近7k行，可以预见，将来不会更短，只会更长。此外，json 这个格式不能很好地支持注释（现在的注释似乎放在一个字段里），线性的文本格式不符合流程的组织形式，这导致了修改和新增流程困难，新开发者上手难度很高。 

我认为这样的组织形式有些不合理，并提出一些改进方案的设想：
1. 彻底重写对 `resource/tasks.json` 的读取方式，将其拆分为若干子文件夹的树状结构，将目录作为 `namespace` 使用。
2. 不改动对 `resource/tasks.json` 的读取方式，将其拆分为若干子文件夹的树状结构，将目录作为 `namespace` 使用，并在发布时将所有子文件编译为单个 `resource/tasks.json`。
3. 不改动对 `resource/tasks.json` 的读取方式，采用某种可视化的编辑方式（如流程图）。

不知道作者怎么看？

**Comment by MistEO** [comment]
拆本身好拆，建个 task 文件夹，读取里面所有的 json 文件 再合到一起。不过现在很多 task 的 next 交织在一起，太乱了，工作量很大，~~主要是没有人手弄~~

做个 gui 工具用来管理似乎是更好的方式，~~但还是没人弄~~

**Reply by crazysmile-PhD** [reply]
gui 工具(能力有限，怕打出屎山，可以試試)後勤加油吧修錯誤和修失效，挺多的，有問題跟我說說我再改改(*/ω＼*)

**Comment by zzyyyl** [comment]
@status102

**Comment by crazysmile-PhD** [comment]
@MistEO @zzyyyl 檔案中儲存的報告包含以下部分：

缺失的參考資料：

本部分列出了所有任務，這些任務的引用指向當前任務集中不存在的任務。這些可能是需要手動更正的損壞引用。next
循環參考：
[task_reference_report.json](https://github.com/user-attachments/files/17148319/task_reference_report.json)

本節包括引用自身或通過將引用點返回到同一任務來創建循環的任務。這些可能會導致執行迴圈，應進行審查以進行可能的調整。next
有效參考：

作為參考，本部分列出了具有有效引用的任務，其中鏈接的任務存在且未檢測到問題。next
該報告以 JSON 格式構建，其中每個部分都清楚地列出，以便於識別和更正。

如果您需要進一步的解釋或其他功能來説明查看報告，請告訴我。查了快一個月，修正方案會給上╰(*°▽°*)╯

**Comment by crazysmile-PhD** [comment]
@MistEO 確定關鍵問題的優先順序：具有最多連接或被大量引用的任務將在清單中排名靠前，以供審核。
提供修復建議：對於每個問
[final_task_reference_report_from_upload.json](https://github.com/user-attachments/files/17148633/final_task_reference_report_from_upload.json)

題，我將根據現有任務引用和一般任務流模式提出修復建議。

**Comment by Alan-Charred** [comment]
支持！tasks 这块儿有相当一部分我看不下去的地方，事实上每次写 tasks 我都感觉有更好的改进空间。为此我正在准备一个关于任务流规范化的讨论 #10679

目前还在忙毕业论文所以没什么时间写，对于任务流的可读性与可视化我有一点思考：

1. 像常见的 concurrency modelling language 那样我将单个任务抽象为 guard->action，因为存在"guard 满足 但是 action点下去无效，需要再点一次" 的问题，除了设置 postDelay 之外，很多任务会选择在 next 中引用自身。我以前会将这个称为回滚，但最近想想，应该叫做 redo-on-blocked，即在 next 无法执行时再次执行自身。大量混杂这种东西极大地破坏了任务流的可读性，即便可视化后也会一团乱。

2. 一些任务，主要是单步任务，仅仅是作为模版任务存在，它们首先可能根本就是纯虚任务（我参考 C++ 纯虚函数的理念如此称呼），不是有效任务流中的一部分。一个任务可能被加入多个任务的 sub，如何合理的可视化将是一个值得讨论的话题。一个任务会被可视化为多个 transition 吗？据我所知，目前的可视化工具对于 sub 字段的支持是很差的。

这么一想，比起 LTS，将任务流抽象为 Petri nets 或许更合适。（这句话和实际应用无关，是我本人的兴趣。）

**Reply by crazysmile-PhD** [reply]
你提出的关于任务流规范化和可视化的思考非常有趣，尤其是将任务抽象为 guard-action 结构的设想，这在并发建模语言中确实有应用。你提到的 "redo-on-blocked" 机制也很有实际意义，特别是对于那些在 "next" 无法执行时需要重新触发的任务。

关于可视化工具对任务 sub 字段的支持问题，确实是一个难点。如果一个任务属于多个子任务集，如何合理且清晰地展示这些关系是关键挑战。或许可以通过层次化展示或动态展开的方式来提高可读性？

**Reply by crazysmile-PhD** [reply]
你提到的这一切都很有价值，尤其在讨论任务流规范化的时候，可以考虑如何让这些抽象概念更容易被实现和理解。希望你的毕业论文顺利完成！

**Comment by MistEO** [comment]
有兴趣可以看下我们新的框架 [MaaFramework](https://github.com/MaaXYZ/MaaFramework)

**Reply by crazysmile-PhD** [reply]
看過了.......什麼意思是在挖角我嗎? Σ(ﾟДﾟ；≡；ﾟдﾟ)(誤

**Reply by crazysmile-PhD** [reply]
@MistEO 我詢問一下(也有可能是我沒看到)有沒有釋出訓練好的神經網路模型嗎?還是要自己DIY  PS因為我在MMA檔案裡面只看到一堆腳本(つд⊂)，找不到有關神經網路調適的內容(◞‸◟)

**Reply by crazysmile-PhD** [reply]
@MistEO 有否其他開發群，在非中國大陸地區是無法使用QQ的(｡･ω･｡)
